File: ios/RNHelloDoctor/Video/HDVideoModule.m
//
//  VideoModule.m
//  hellodoctor
//
//  Created by HelloDoctor on 3/21/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>

#import "HDVideoModule.h"

@implementation HDVideoModule
//
//+ (BOOL)requiresMainQueueSetup
//{
//    return YES;
//}
//

RCT_EXPORT_MODULE(HDVideo)

RCT_REMAP_METHOD(connect,
                 connectToRoom:(NSString*)roomName
                 withAccessToken:(NSString*)token
                 connectToRoomResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo connectWithRoomName:roomName accessToken:token];
    
    resolve(@"connected");
}

RCT_REMAP_METHOD(isConnectedToRoom,
                 isConnectToRoom:(NSString*)roomName
                 isConnectedResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    BOOL isConnected = [hdVideo isConnectedToRoomName:roomName];
    
    resolve([NSString stringWithFormat:@"%d", isConnected]);
}

RCT_REMAP_METHOD(getRemoteParticipants,
                 getRemoteParticipants:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    NSArray* remoteParticipants = [hdVideo getRemoteParticipants];
    
    resolve(remoteParticipants);
}

RCT_REMAP_METHOD(disconnect,
                 disconnectResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo disconnect];
    
    resolve(@"disconnected");
}

RCT_REMAP_METHOD(setVideoPublished,
                 isPublished:(BOOL)published
                 enableVideoResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo setLocalVideoPublishedWithPublished:published];
    
    resolve([NSString stringWithFormat:@"local video published: %d", published]);
}

RCT_REMAP_METHOD(setVideoEnabled,
                 isEnabled:(BOOL)enabled
                 enableVideoResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo setLocalVideoEnabledWithEnabled:enabled];
    
    resolve([NSString stringWithFormat:@"local video enabled: %d", enabled]);
}

RCT_REMAP_METHOD(setAudioEnabled,
                 isEnabled:(BOOL)enabled
                 enableAudioResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo setLocalAudioEnabledWithEnabled:enabled];
    
    resolve([NSString stringWithFormat:@"local audio enabled: %d", enabled]);
}

RCT_REMAP_METHOD(setSpeakerPhone,
                 withSpeakerEnabled:(BOOL)enabled
                 setSpeakerPhoneResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo setSpeakerEnabledWithEnabled:enabled];
    
    resolve(@"success");
}

RCT_REMAP_METHOD(flipCamera,
                 flipCameraResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    HDVideo* hdVideo = [HDVideo getInstance];
    
    [hdVideo flipCamera];
    
    resolve(@"success");
}

@end

File: ios/RNHelloDoctor/Video/HDVideoModule.swift
//
//  HDVideoModule.swift
//  RNHelloDoctor
//
//  Created by HelloDoctor on 3/29/22.
//

import Foundation

@objc(HDVideoModule)
class HDVideoModule: RCTEventEmitter {
    @objc(connect:withAccessToken:withResolver:withRejecter:)
    func connect(roomName: String, accessToken: String, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.connect(roomName: roomName, accessToken: accessToken)
        
        resolve("connected")
    }
    
    @objc(isConnectedToRoom:withResolver:withRejecter:)
    func connect(roomName: String, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        
        let isConnected = hdVideo.isConnectedTo(roomName: roomName)
        
        resolve("\(isConnected)")
    }
    
    @objc(getRemoteParticipants:withRejecter:)
    func getRemoteParticipants(resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        
        let remoteParticipants = hdVideo.getRemoteParticipants()
        
        resolve(remoteParticipants)
    }
    
    @objc(disconnect:withRejecter:)
    func disconnect(resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.disconnect()
        
        resolve("disconnected")
    }
    
    @objc(setVideoPublished:withResolver:withRejecter:)
    func setVideoPublished(isPublished: Bool, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.setLocalVideoPublished(published: isPublished)
        
        resolve("set local video published: \(isPublished)")
    }
    
    @objc(setVideoEnabled:withResolver:withRejecter:)
    func setVideoEnabled(isEnabled: Bool, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.setLocalVideoEnabled(enabled: isEnabled)
        
        resolve("set local video enabled: \(isEnabled)")
    }
    
    @objc(setAudioEnabled:withResolver:withRejecter:)
    func setAudioEnabled(isEnabled: Bool, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.setLocalAudioEnabled(enabled: isEnabled)
        
        resolve("set local audio enabled: \(isEnabled)")
    }
    
    @objc(setSpeakerPhone:withResolver:withRejecter:)
    func setSpeakerPhone(isEnabled: Bool, resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.setSpeakerEnabled(enabled: isEnabled)
        
        resolve("set speaker enabled: \(isEnabled)")
    }
    
    @objc(flipCamera:withRejecter:)
    func flipCamera(resolve:@escaping RCTPromiseResolveBlock, reject:@escaping RCTPromiseRejectBlock) -> Void {
        let hdVideo = HDVideo.getInstance()
        hdVideo.flipCamera()
        
        resolve("flipped camera")
    }
}
